libpromises/dbm_api.c

```diff
@@ -180,6 +253,40 @@ static DBHandle *DBHandleGet(int id)
     return &db_handles[id];
 }
 
+static inline
+void CloseDBInstance(DBHandle *handle)
+{
+    /* Wait until all DB users are served, or a threshold is reached */
+    int count = 0;
+    ThreadLock(&handle->lock);
+    while (handle->refcount > 0 && count < 1000) {
+        ThreadUnlock(&handle->lock);
+
+        struct timespec sleeptime = {
+            .tv_sec = 0,
+            .tv_nsec = 10000000 /* 10 ms */
+        };
+        nanosleep(&sleeptime, NULL);
+        count++;
+
+        ThreadLock(&handle->lock);
+    }
+    /* Keep mutex locked. */
+
+    /* If we exited because of timeout make sure we Log() it. */
+    if (handle->refcount != 0) {
+        Log(LOG_LEVEL_ERR,
+                "Database %s refcount is still not zero (%d), forcing CloseDB()!",
+                handle->filename, handle->refcount);
+        DBPrivCloseDB(handle->priv);
+    } else /* TODO: can we clean this up unconditionally ? */ {
```

Comment:
Fixed! My editor has option to beautify the code after copy-paste which is not perfect all the time...